### PR TITLE
Add OpenCL language support

### PR DESCRIPTION
**Pull request by Milania1**
This PR adds support for the [OpenCL](https://www.khronos.org/opencl/) language. It adds the new `language-opencl` class to support OpenCL kernel code and additionally extends the existing c and cpp languages to highlight some of the OpenCL host code.

I already use it on my own site, e.g. in [this article](https://milania.de/blog/Performance_evaluation_of_image_convolution_with_gradient_filters_in_OpenCL).

### CODE REVIEW

File: components/prism-opencl.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
...ble8|double8x1|double8x16|double8x2|double8x4|double8x8|else|enum|event_t|extern|float|float16|float16x1|float16x16|float16x2|float16x4|float16x8|float1x1|float1x16|float1x2|float1x4|float1x8|float2|float2x1|float2x16|float2x2|float2x4|float2x8|float3|float4|float4x1|float4x16|float4x2|float4x4|float4x8|float8|float8x1|float8x16|float8x2|float8x4|float8x8|for|global|goto|half|half16|half2|half3|half4|half8|if|image1d_array_t|image1d_buffer_t|image1d_t|image2d_array_depth_t|image2d_array_msaa_depth_t|image2d_array_msaa_t|image2d_array_t|image2d_depth_t|image2d_msaa_depth_t|image2d_msaa_t|image2d_t|image3d_t|imaginary|inline|int|int16|int2|int3|int4|int8|intptr_t|kernel|local|long|long16|long2|long3|long4|long8|ndrange_t|packed|pipe|private|ptrdiff_t|quad|quad16|quad2|quad3|quad4|quad8|queue_t|read_only|read_write|register|reserve_id_t|restrict|return|sampler_t|short|short16|short2|short3|short4|short8|signed|size_t|sizeof|static|struct|switch|typedef|uchar|uchar16|uchar2|uchar3|uchar4|uchar8|uint|uint16|uint2|uint3|uint4|uint8|uintptr_t|ulong|ulong16|ulong2|ulong3|ulong4|ulong8|uniform|union|unsigned|ushort|ushort16|ushort2|ushort3|ushort4|ushort8|void|volatile|while|write_only)\b/,
```
Comment (Golmote) [inline components/prism-opencl.js]: Wow those are huge regexps. You might want to consider at least combining some of them?

-  `bool(16|2|3|4|8)?`
- `double(16(x(1|16|2|4|8))?|1x(1|16|2|4|8)|2(x(1|16|2|4|8))?|3|4(x(1|16|2|4|8))?|8(x(1|16|2|4|8))?)?`
- `image(1d_(array_|buffer_)?t|2d_(array_(depth_|msaa_depth_|msaa_)?|depth_|msaa_depth_|msaa_)?t|3d_t)`
etc.

File: components/prism-opencl.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
...ada_half8_rtz|vstore16|vstore2|vstore3|vstore4|vstore8|vstore_half|vstore_half16|vstore_half16_rte|vstore_half16_rtn|vstore_half16_rtp|vstore_half16_rtz|vstore_half2|vstore_half2_rte|vstore_half2_rtn|vstore_half2_rtp|vstore_half2_rtz|vstore_half3|vstore_half3_rte|vstore_half3_rtn|vstore_half3_rtp|vstore_half3_rtz|vstore_half4|vstore_half4_rte|vstore_half4_rtn|vstore_half4_rtp|vstore_half4_rtz|vstore_half8|vstore_half8_rte|vstore_half8_rtn|vstore_half8_rtp|vstore_half8_rtz|vstore_half_rte|vstore_half_rtn|vstore_half_rtp|vstore_half_rtz|vstore_rte|vstore_rtn|vstore_rtp|vstore_rtz|vstorea_half16|vstorea_half16_rte|vstorea_half16_rtn|vstorea_half16_rtp|vstorea_half16_rtz|vstorea_half2|vstorea_half2_rte|vstorea_half2_rtn|vstorea_half2_rtp|vstorea_half2_rtz|vstorea_half3|vstorea_half3_rte|vstorea_half3_rtn|vstorea_half3_rtp|vstorea_half3_rtz|vstorea_half4|vstorea_half4_rte|vstorea_half4_rtn|vstorea_half4_rtp|vstorea_half4_rtz|vstorea_half8|vstorea_half8_rte|vstorea_half8_rtn|vstorea_half8_rtp|vstorea_half8_rtz|vstorea_half_rte|vstorea_half_rtn|vstorea_half_rtp|vstorea_half_rtz|wait_group_events|work_group_size_hint|write_imagef|write_imageh|write_imagei|write_imageui|write_mem_fence)\b/,
```
Comment (Golmote) [inline components/prism-opencl.js]: Same comment here. I can see a lot of common prefixes that could help combining those alternatives into a shorter regexp. (`native_`, `read_`, `vstore_`, etc.)

File: components/prism-opencl.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+			alias: 'function'
+		},
+		// Extracted from http://streamcomputing.eu/downloads/?opencl.lang (opencl-const)
+		'constant-opencl-kernel': {
+			pattern: /\b(CHAR_BIT|CHAR_MAX|CHAR_MIN|CLK_ADDRESS_CLAMP|CLK_ADDRESS_CLAMP_TO_EDGE|CLK_ADDRESS_NONE|CLK_ADDRESS_REPEAT|CLK_FILTER_LINEAR|CLK_FILTER_NEAREST|CLK_GLOBAL_MEM_FENCE|CLK_LOCAL_MEM_FENCE|CLK_NORMALIZED_COORDS_FALSE|CLK_NORMALIZED_COORDS_TRUE|CL_A|CL_ARGB|CL_BGRA|CL_FLOAT|CL_HALF_FLOAT|CL_INTENSITY|CL_LUMINANCE|CL_R|CL_RA|CL_RG|CL_RGB|CL_RGBA|CL_RGBx|CL_RGx|CL_Rx|CL_SIGNED_INT16|CL_SIGNED_INT32|CL_SIGNED_INT8|CL_SNORM_INT16|CL_SNORM_INT8|CL_UNORM_INT16|CL_UNORM_INT8|CL_UNORM_INT_101010|CL_UNORM_SHORT_555|CL_UNORM_SHORT_565|CL_UNSIGNED_INT16|CL_UNSIGNED_INT32|CL_UNSIGNED_INT8|DBL_DIG|DBL_EPSILON|DBL_MANT_DIG|DBL_MAX|DBL_MAX_10_EXP|DBL_MIN|DBL_MIN_10_EXP|DBL_MIN_EXP|FLT_DIG|FLT_EPSILON|FLT_MANT_DIG|FLT_MAX|FLT_MAX_10_EXP|FLT_MAX_EXP|FLT_MIN|FLT_MIN_10_EXP|FLT_MIN_EXP|FLT_RADIX|HUGE_VALF|INFINITY|INT_MAX|INT_MIN|LONG_MAX|LONG_MIN|MAXFLOAT|M_1_PI|M_2_PI|M_2_SQRTPI|M_E|M_LN10|M_LN2|M_LOG10E|M_LOG2E|M_PI|M_PI2|M_PI4|M_SQRT1_2|M_SQRT2|NAN|SCHAR_MAX|SCHAR_MIN|SHRT_MAX|SHRT_MIN|UCHAR_MAX|UCHAR_MIN|UINT_MAX|ULONG_MAX)\b/,
```
Comment (Golmote) [inline components/prism-opencl.js]: Same comment here.

File: components/prism-opencl.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
...status|cl_channel_order|cl_channel_type|cl_char|cl_char16|cl_char2|cl_char3|cl_char4|cl_char8|cl_command_queue|cl_command_queue_info|cl_command_queue_properties|cl_command_type|cl_context|cl_context_info|cl_context_properties|cl_device_exec_capabilities|cl_device_fp_config|cl_device_id|cl_device_info|cl_device_local_mem_type|cl_device_mem_cache_type|cl_device_type|cl_double|cl_double16|cl_double2|cl_double3|cl_double4|cl_double8|cl_event|cl_event_info|cl_filter_mode|cl_float|cl_float16|cl_float2|cl_float3|cl_float4|cl_float8|cl_half|cl_image_info|cl_int|cl_int16|cl_int2|cl_int3|cl_int4|cl_int8|cl_kernel|cl_kernel_info|cl_kernel_work_group_info|cl_long|cl_long16|cl_long2|cl_long3|cl_long4|cl_long8|cl_map_flags|cl_mem|cl_mem_flags|cl_mem_info|cl_mem_object_type|cl_platform_id|cl_platform_info|cl_profiling_info|cl_program|cl_program_build_info|cl_program_info|cl_sampler|cl_sampler_info|cl_short|cl_short16|cl_short2|cl_short3|cl_short4|cl_short8|cl_uchar|cl_uchar16|cl_uchar2|cl_uchar3|cl_uchar4|cl_uchar8|cl_uint|cl_uint16|cl_uint2|cl_uint3|cl_uint4|cl_uint8|cl_ulong|cl_ulong16|cl_ulong2|cl_ulong3|cl_ulong4|cl_ulong8|cl_ushort|cl_ushort16|cl_ushort2|cl_ushort3|cl_ushort4|cl_ushort8)\b/,
```
Comment (Golmote) [inline components/prism-opencl.js]: Same comment here. All of those share the `cl_` prefix!

File: components/prism-opencl.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
..._PROGRAM_BUILD_GLOBAL_VARIABLE_TOTAL_SIZE|CL_PROGRAM_BUILD_LOG|CL_PROGRAM_BUILD_OPTIONS|CL_PROGRAM_BUILD_STATUS|CL_PROGRAM_CONTEXT|CL_PROGRAM_DEVICES|CL_PROGRAM_IL|CL_PROGRAM_KERNEL_NAMES|CL_PROGRAM_NUM_DEVICES|CL_PROGRAM_NUM_KERNELS|CL_PROGRAM_REFERENCE_COUNT|CL_PROGRAM_SOURCE|CL_QUEUED|CL_QUEUE_CONTEXT|CL_QUEUE_DEVICE|CL_QUEUE_DEVICE_DEFAULT|CL_QUEUE_ON_DEVICE|CL_QUEUE_ON_DEVICE_DEFAULT|CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE|CL_QUEUE_PROFILING_ENABLE|CL_QUEUE_PROPERTIES|CL_QUEUE_REFERENCE_COUNT|CL_QUEUE_SIZE|CL_R|CL_RA|CL_READ_ONLY_CACHE|CL_READ_WRITE_CACHE|CL_RG|CL_RGB|CL_RGBA|CL_RGBx|CL_RGx|CL_RUNNING|CL_Rx|CL_SAMPLER_ADDRESSING_MODE|CL_SAMPLER_CONTEXT|CL_SAMPLER_FILTER_MODE|CL_SAMPLER_LOD_MAX|CL_SAMPLER_LOD_MIN|CL_SAMPLER_MIP_FILTER_MODE|CL_SAMPLER_NORMALIZED_COORDS|CL_SAMPLER_REFERENCE_COUNT|CL_SIGNED_INT16|CL_SIGNED_INT32|CL_SIGNED_INT8|CL_SNORM_INT16|CL_SNORM_INT8|CL_SUBMITTED|CL_SUCCESS|CL_UNORM_INT16|CL_UNORM_INT24|CL_UNORM_INT8|CL_UNORM_INT_101010|CL_UNORM_INT_101010_2|CL_UNORM_SHORT_555|CL_UNORM_SHORT_565|CL_UNSIGNED_INT16|CL_UNSIGNED_INT32|CL_UNSIGNED_INT8|CL_VERSION_1_0|CL_VERSION_1_1|CL_VERSION_1_2|CL_VERSION_2_0|CL_VERSION_2_1|CL_sBGRA|CL_sRGB|CL_sRGBA|CL_sRGBx)\b/,
```
Comment (Golmote) [inline components/prism-opencl.js]: Same comment here. They seem to all share the `CL_` prefix, plus other parts are ofter repeated.

File: components/prism-opencl.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
...MMigrateMem|clEnqueueSVMUnmap|clEnqueueTask|clEnqueueUnmapMemObject|clEnqueueWaitForEvents|clEnqueueWriteBuffer|clEnqueueWriteBufferRect|clEnqueueWriteImage|clFinish|clFlush|clGetCommandQueueInfo|clGetContextInfo|clGetDeviceAndHostTimer|clGetDeviceIDs|clGetDeviceInfo|clGetEventInfo|clGetEventProfilingInfo|clGetExtensionFunctionAddress|clGetExtensionFunctionAddressForPlatform|clGetHostTimer|clGetImageInfo|clGetKernelArgInfo|clGetKernelInfo|clGetKernelSubGroupInfo|clGetKernelWorkGroupInfo|clGetMemObjectInfo|clGetPipeInfo|clGetPlatformIDs|clGetPlatformInfo|clGetProgramBuildInfo|clGetProgramInfo|clGetSamplerInfo|clGetSupportedImageFormats|clLinkProgram|clReleaseCommandQueue|clReleaseContext|clReleaseDevice|clReleaseEvent|clReleaseKernel|clReleaseMemObject|clReleaseProgram|clReleaseSampler|clRetainCommandQueue|clRetainContext|clRetainDevice|clRetainEvent|clRetainKernel|clRetainMemObject|clRetainProgram|clRetainSampler|clSVMAlloc|clSVMFree|clSetCommandQueueProperty|clSetDefaultDeviceCommandQueue|clSetEventCallback|clSetKernelArg|clSetKernelArgSVMPointer|clSetKernelExecInfo|clSetMemObjectDestructorCallback|clSetUserEventStatus|clUnloadCompiler|clUnloadPlatformCompiler|clWaitForEvents)\b/,
```
Comment (Golmote) [inline components/prism-opencl.js]: Same comment here. `cl` prefix plus other common parts.

**Comment by Milania1** [comment]
Thanks for the review!

I updated the regexes so that they are (a bit) shorter.

**Comment by Golmote** [comment]
Merged! Thank you for contributing!

**Pull request merged**